:book: Add pivoting documentation to user-guide (#254)
metal3-io/metal3-docs: docs/user-guide/src/capm3/pivoting.md

# CAPM3 Pivoting

## What is pivoting

Cluster-api-provider-metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.

CAPI Pivoting feature is a process of moving the provider components and declared Cluster API resources from a source
management cluster to a target management cluster by using the `clusterctl` functionality called "move".
More information about the general CAPI "move" functionality can be find [here](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html).

In Metal3, pivoting is performed by using the CAPI `clusterctl` tool provided by Cluster-API. `clusterctl` recognizes pivoting as move.
During the pivot process `clusterctl` pauses any reconciliation of Cluster-API objects and this gets propagated to (CAPM3) objects as well.
Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the

> source cluster.

Reviewer: Is the deletion always true ? and do we do it that way ?
@kashifest

Reviewer 2: Yes

## Prerequisits

* **It is mandatory to use `clusterctl` for both the bootstrap and target cluster.**

  If the provider components are not installed using `clusterctl`, it will not
  be able to identify the objects to move.  Initializing the cluster using

>   `clusterctl` essentially adds the following labels in the CRDs of each object
>   related.

Reviewer: ```suggestion
  `clusterctl` essentially adds the following labels in the CRDs of each related object.
```

Author: done.

  ```yaml
  labels:
  - clusterctl.cluster.x-k8s.io: ""
  - cluster.x-k8s.io/provider: "<provider-name>"
  ```

  So if the clusters are not initialized using `clusterctl`, all the CRDS of the
  objects to be moved to target cluster needs to have these labels both in

>   bootstrap cluster and target clusteri befor performing the move. **Note**: This is not recommended, since

Reviewer: ```suggestion
  bootstrap cluster and target cluster before performing the move. **Note**: This is not recommended, since
```

Author: done

>   the way `clusterctl` identifies objects to manage might change in future, so

Reviewer: ```suggestion
  the way `clusterctl` identifies objects to manage might change in the future, so
```

Author: done

>   it's always safe to install CRDs and controllers through `clusterctl` init.

Reviewer: ```suggestion
  it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.
```

Author: done

* **BMH objects have correct status annotation.**

  Since BMH _status_ holds important information regarding the BMH itself, BMH with
  status has to be moved and BMH has to be reconstructed with correct status in
  target  cluster before it is being reconciled. This is now done through BMH
  status annotation in BMO.

* **Maintain connectivity towards provisioning network.**

  Baremetal machines boot over a network with a DHCP server. This requires
  maintaining a fixed IP end points towards the provisioning network. This is
  achieved through _keepalived_. A new container is added namely
  _ironic-endpoint-keepalived_ which maintains the Ironic Endpoint using
  keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
  is to ensure that the Ironic Endpoint IP is also passed onto the target
  cluster control plane. This also guarantees that once moving is done and the
   management cluster is taken down, target cluster controlplane can re-claim

>    the ironic endpoint IP through keepalived. The end goal is to make ironic

Reviewer: ```suggestion
   the Ironic endpoint IP through keepalived. The end goal is to make Ironic
```

Author: done

* **BMO is deployed as part of CAPM3.**

  If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
  need to be labeled accordingly manually. This is a prerequisite for both the
  management and the target cluster.

* **Objects should have a proper owner reference chain.**

  `clusterctl move` moves all the objects to the target cluster following the
[owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
So, it is necessary to verify that all the desired objects that needs to
be moved to the target cluster have a proper owner reference chain.

## Step by step pivoting process

As described in
[clusterctl](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html)
the whole process of bootstrapping a management cluster to moving
objects to target cluster can be described as follows:

The move process can be bounded with the creation of a temporary bootstrap
cluster used to provision a target Management cluster.

This can now be achieved with the following procedure:

1. Create a temporary bootstrap cluster, e.g. using Kind or Minikube using
 `clusterctl`.

> 2. Install ironic  components namely: ironic, ironic-inspector, ironic-database,

Reviewer: ```suggestion
2. Install Ironic  components namely: ironic, ironic-inspector, ironic-database,
```

Author: done

 on namespace `metal3` and they will be watching over objects in namespace `metal3`.

4. Provision target cluster:

   Example:

   `clusterctl config cluster ... | kubectl apply -f -`

5. Wait for the target management cluster to be up and running and once it is up
 get the kubeconfig for the new target management cluster.

6. Use the new cluster’s kubeconfig to install the ironic-components in the
 target cluster.

7. Use `clusterctl` init with the new cluster’s kubeconfig to install the provider
 components.

    Example:

    `clusterctl init --kubeconfig target.yaml --infrastructure metal3:v0.3.1
  --target-namespace metal3 --watching-namespace metal3`

8. Use `clusterctl` move to move the Cluster API resources from the bootstrap
 cluster to the target management cluster.

    Example:

    `clusterctl move --to-kubeconfig target.yaml -n metal3 -v 10`

9. Delete the bootstrap cluster

## Important Notes

The following requirements are essential for the move process to run
successfully:

1. The move process should be done when the BMHs are in a steady state. BMHs
   should not be moved while any operation is on-going i.e. BMH is in provisioning

>    state. This will result in failure since the interaction between IPA and ironic

Reviewer: ```suggestion
   state. This will result in failure since the interaction between IPA and Ironic
```

Author: done

>    gets broken and as a result ironic's database might not be repopulated and

Reviewer: ```suggestion
   gets broken and as a result Ironic's database might not be repopulated and
```

Author: done

   eventually the cluster will end up in an erroneous state. Moreover, the IP of
   the BMH might change after the move and the DHCP-leases from the management
   cluster are not moved to target cluster.

> 2. Before the move process is initialized, it is important to delete the ironic
>    pod/ironic containers. If ironic is deployed in cluster the deployment is named
>    `metal3-ironic`, if it is deployed locally outside the cluster, the containers
>    to be deleted are `ironic`, `ironic-inspector`, `ironic-endpoint-keepalived`,
>    `ironic-dnsmasq`, `ironic-httpd` and `mariadb`. If ironic is not deleted
>    before move, the old ironic might interfere with the operations of
>    the new ironic deployed in target cluster since the database of the first ironic
>    instance is not cleaned when the BMHs are moved. Also there would be two dnsmasq
>    existent in the deployment if there would be two ironic deployment which is
>    undesirable.

Reviewer: ```suggestion
2. Before the move process is initialized, it is important to delete the Ironic
   pod/ironic containers. If Ironic is deployed in cluster the deployment is named
   `metal3-ironic`, if it is deployed locally outside the cluster, the containers
   to be deleted are `ironic`, `ironic-inspector`, `ironic-endpoint-keepalived`,
   `ironic-dnsmasq`, `ironic-httpd` and `mariadb`. If Ironic is not deleted
   before move, the old ironic might interfere with the operations of
   the new ironic deployed in target cluster since the database of the first Ironic
   instance is not cleaned when the BMHs are moved. Also there would be two dnsmasq
   existent in the deployment if there would be two ironic deployment which is
   undesirable.
```

Author: done

3. The provisioning bridge where the `ironic-endpoint-IP` is supposed to be

>    attached should have a static IP assignment on it before the ironic

Reviewer: ```suggestion
   attached to should have a static IP assignment on it before the ironic
```

Author: done

   pod/containers start to operate in the target cluster. This is important since
   `ironic-endpoint-keepalived` container will only assign the `ironic-endpoint-IP`
   on the provisioning bridge in target cluster when it has an IP on it. Otherwise

>    it will fail to attach the IP and ironic will be unreachable. This is crucial

Reviewer: ```suggestion
   it will fail to attach the IP and Ironic will be unreachable. This is crucial
```

Author: done

# CAPM3 Pivoting

## What is pivoting

> Cluster-api-provider-metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.

Reviewer 3: ```suggestion
Cluster API Provider Metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.
```

CAPI Pivoting feature is a process of moving the provider components and declared Cluster API resources from a source
management cluster to a target management cluster by using the `clusterctl` functionality called "move".

> More information about the general CAPI "move" functionality can be find [here](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html).

Reviewer 2: ```suggestion
More information about the general CAPI "move" functionality can be found [here](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html).
```

In Metal3, pivoting is performed by using the CAPI `clusterctl` tool provided by Cluster-API. `clusterctl` recognizes pivoting as move.
During the pivot process `clusterctl` pauses any reconciliation of Cluster-API objects and this gets propagated to (CAPM3) objects as well.
Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
source cluster.

> ## Prerequisits

Reviewer 3: ```suggestion
## Prerequisite
```

* **It is mandatory to use `clusterctl` for both the bootstrap and target cluster.**

  If the provider components are not installed using `clusterctl`, it will not
  be able to identify the objects to move.  Initializing the cluster using
  `clusterctl` essentially adds the following labels in the CRDs of each related object.

  ```yaml
  labels:
  - clusterctl.cluster.x-k8s.io: ""
  - cluster.x-k8s.io/provider: "<provider-name>"
  ```

  So if the clusters are not initialized using `clusterctl`, all the CRDS of the
  objects to be moved to target cluster needs to have these labels both in
  bootstrap cluster and target cluster before performing the move.

  **Note**: This is not recommended, since
  the way `clusterctl` identifies objects to manage might change in the future, so
  it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.

* **BMH objects have correct status annotation.**

  Since BMH _status_ holds important information regarding the BMH itself, BMH with

>   status has to be moved and BMH has to be reconstructed with correct status in

Reviewer 3: ```suggestion
  status has to be moved and it has to be reconstructed with correct status in
```

  target  cluster before it is being reconciled. This is now done through BMH

>   status annotation in BMO.

Reviewer 3: you could provide a reference to BMO doc here

Author: I would rather have somebody add it later when the bmo documentation is part of the user-guide.

Reviewer 3: Is there an open issue to add it to the user guide? If not, why not to reference it

Author: Yes there is https://github.com/metal3-io/metal3-docs/issues/221, but based on the PR https://github.com/metal3-io/metal3-docs/pull/237 I can preemptively place the link, I hope the file names won't change.

* **Maintain connectivity towards provisioning network.**

  Baremetal machines boot over a network with a DHCP server. This requires
  maintaining a fixed IP end points towards the provisioning network. This is
  achieved through _keepalived_. A new container is added namely

>   _ironic-endpoint-keepalived_ which maintains the Ironic Endpoint using

Reviewer 2: ```suggestion
  _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
```

  keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
  is to ensure that the Ironic Endpoint IP is also passed onto the target
  cluster control plane. This also guarantees that once moving is done and the
   management cluster is taken down, target cluster controlplane can re-claim
   the Ironic endpoint IP through keepalived. The end goal is to make Ironic

>    endpoint reachable in the target cluster.

Reviewer 3: these lines shifted to the right for some reason

* **BMO is deployed as part of CAPM3.**

  If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
  need to be labeled accordingly manually. This is a prerequisite for both the

>   management and the target cluster.

Reviewer 3: You could mention the CAPM3 API version which deploys BMO as part of it and the ones does not to make it more clear for the user

* **Objects should have a proper owner reference chain.**

  `clusterctl move` moves all the objects to the target cluster following the
[owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
So, it is necessary to verify that all the desired objects that needs to
be moved to the target cluster have a proper owner reference chain.

## Step by step pivoting process

As described in
[clusterctl](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html)
the whole process of bootstrapping a management cluster to moving
objects to target cluster can be described as follows:

The move process can be bounded with the creation of a temporary bootstrap

> cluster used to provision a target Management cluster.

Reviewer 3: ```suggestion
cluster used to provision a target management cluster.
```

This can now be achieved with the following procedure:

1. Create a temporary bootstrap cluster, e.g. using Kind or Minikube using

>  `clusterctl`.

Reviewer 2: kind or minikube is not setup using clusterctl. Rather the provider components in kind or minikube are. This line needs to be tweaked a little bit to phrase this appropriately.

2. Install Ironic  components namely: ironic, ironic-inspector, ironic-database,

>    ironic-endpoint-keepalived, httpd and dnsmasq.

Reviewer 2: ironic-database? do we have a container with that name ?

3. Use clusterctl init to install the provider components

   Example:

>    `clusterctl init --infrastructure metal3:v1.1.1

Reviewer 3: we do not have the CAPM3 v1.1.1 release yet

   --target-namespace metal3 --watching-namespace metal3`

 This command will create the necessary CAPI controllers (CAPI, CABPK, CAKCP)
 and CAPM3 as the infrastructure provider. All of the controllers will be installed
 on namespace `metal3` and they will be watching over objects in namespace `metal3`.

4. Provision target cluster:

   Example:

   `clusterctl config cluster ... | kubectl apply -f -`

5. Wait for the target management cluster to be up and running and once it is up
 get the kubeconfig for the new target management cluster.

6. Use the new cluster’s kubeconfig to install the ironic-components in the
 target cluster.

7. Use `clusterctl` init with the new cluster’s kubeconfig to install the provider
 components.

    Example:

>     `clusterctl init --kubeconfig target.yaml --infrastructure metal3:v0.3.1

Reviewer 2: Can you keep the same version for metal3 in the example as in step 3

8. Use `clusterctl` move to move the Cluster API resources from the bootstrap
 cluster to the target management cluster.

    Example:

    `clusterctl move --to-kubeconfig target.yaml -n metal3 -v 10`

9. Delete the bootstrap cluster

## Important Notes

The following requirements are essential for the move process to run
successfully:

1. The move process should be done when the BMHs are in a steady state. BMHs
   should not be moved while any operation is on-going i.e. BMH is in provisioning
   state. This will result in failure since the interaction between IPA and Ironic
   gets broken and as a result Ironic's database might not be repopulated and
   eventually the cluster will end up in an erroneous state. Moreover, the IP of
   the BMH might change after the move and the DHCP-leases from the management
   cluster are not moved to target cluster.

2. Before the move process is initialized, it is important to delete the Ironic
   pod/Ironic containers. If Ironic is deployed in cluster the deployment is named
   `metal3-ironic`, if it is deployed locally outside the cluster, the containers
   to be deleted are `ironic`, `ironic-inspector`, `ironic-endpoint-keepalived`,

>    `ironic-dnsmasq`, `ironic-httpd` and `mariadb`. If Ironic is not deleted

Reviewer 2: is mariadb still deployed locally?

Author: I think I will remove the text related to running ironic locally as that is totally up to the user how the ironic is configured in that case.

# CAPM3 Pivoting

## What is pivoting

> Cluster Api Provider Metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.

Reviewer 4: ```suggestion
Cluster API Provider Metal3 (CAPM3) implements support for Cluster API (CAPI)'s 'move/pivoting' feature.
```

During the pivot process `clusterctl` pauses any reconciliation of Cluster-API objects and this gets propagated to (CAPM3) objects as well.
Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
source cluster.

## Prerequisite

* **It is mandatory to use `clusterctl` for both the bootstrap and target cluster.**

  If the provider components are not installed using `clusterctl`, it will not
  be able to identify the objects to move.  Initializing the cluster using
  `clusterctl` essentially adds the following labels in the CRDs of each related object.

  ```yaml
  labels:
  - clusterctl.cluster.x-k8s.io: ""
  - cluster.x-k8s.io/provider: "<provider-name>"
  ```

  So if the clusters are not initialized using `clusterctl`, all the CRDS of the
  objects to be moved to target cluster needs to have these labels both in
  bootstrap cluster and target cluster before performing the move.

  **Note**: This is not recommended, since
  the way `clusterctl` identifies objects to manage might change in the future, so
  it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.

> * **BMH objects have correct status annotation.**

Reviewer 4: ```suggestion
* **BareMetalHost objects have correct status annotation.**
```

>   Since BMH _status_ holds important information regarding the BMH itself, BMH with

Reviewer 4: ```suggestion
  Since BareMetalHost (BMH) _status_ holds important information regarding the BMH itself, BMH with
```

  status has to be moved and it has to be reconstructed with correct status in
  target  cluster before it is being reconciled. This is now done through BMH

>   status annotation in [BMO](../bmo/introduction.md).

Reviewer 4: ```suggestion
  [status annotation](https://github.com/metal3-io/baremetal-operator/blob/main/docs/statusAnnotation.md).
```

Author: I think it would be better if we would reference the BMO documentation of the same book, that will also contain examples of the BMH CRDs.

Reviewer 4: agreed

* **Maintain connectivity towards provisioning network.**

  Baremetal machines boot over a network with a DHCP server. This requires
  maintaining a fixed IP end points towards the provisioning network. This is
  achieved through _keepalived_. A new container is added namely
  _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
  keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
  is to ensure that the Ironic Endpoint IP is also passed onto the target
  cluster control plane. This also guarantees that once moving is done and the
  management cluster is taken down, target cluster controlplane can re-claim
  the Ironic endpoint IP through keepalived. The end goal is to make Ironic
  endpoint reachable in the target cluster.

* **BMO is deployed as part of CAPM3.**

>   If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
>   need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
>   because since CAPM3 version [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
>   BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
>   This is a prerequisite for both the management and the target cluster.

Reviewer 4: ```suggestion
  If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
  need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
  because since CAPM3 release [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
  BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
  This is a prerequisite for both the management and the target cluster.
```
Can we also mention what labels need to be applied?

Author: It has been mentioned between line 24 to 28

[owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
So, it is necessary to verify that all the desired objects that needs to
be moved to the target cluster have a proper owner reference chain.

## Step by step pivoting process

As described in
[clusterctl](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html)
the whole process of bootstrapping a management cluster to moving
objects to target cluster can be described as follows:

The move process can be bounded with the creation of a temporary bootstrap
cluster used to provision a target management cluster.

This can now be achieved with the following procedure:

1. Create a temporary bootstrap cluster, the temporary bootstrap cluster (source cluster)
 could be created tools like e.g. using Kind or Minikube using and after the source cluster
 is up and running then the CAPI and provider components can be installed with `clusterctl` to
 the source cluster.

2. Install Ironic  components namely: ironic, ironic-inspector,
   ironic-endpoint-keepalived, httpd and dnsmasq.

3. Use clusterctl init to install the provider components

>    Example:
> 
>    `clusterctl init --infrastructure metal3:v1.1.0
>    --target-namespace metal3 --watching-namespace metal3`

Reviewer 4: ```suggestion
   ```bash
   clusterctl init --infrastructure metal3:v1.1.0
   --target-namespace metal3 --watching-namespace metal3
   ```
```

>  This command will create the necessary CAPI controllers (CAPI, CABPK, CAKCP)
>  and CAPM3 as the infrastructure provider. All of the controllers will be installed
>  on namespace `metal3` and they will be watching over objects in namespace `metal3`.

Reviewer 4: There is indentation issue if check https://deploy-preview-254--metal3io.netlify.app/capm3/pivoting.html
```suggestion
     This command will create the necessary CAPI controllers (CAPI, CABPK, CAKCP)
     and CAPM3 as the infrastructure provider. All of the controllers will be installed
     on namespace `metal3` and they will be watching over objects in namespace `metal3`.
```

4. Provision target cluster:

   Example:

   `clusterctl config cluster ... | kubectl apply -f -`

5. Wait for the target management cluster to be up and running and once it is up
 get the kubeconfig for the new target management cluster.

6. Use the new cluster’s kubeconfig to install the ironic-components in the
 target cluster.

> 7. Use `clusterctl` init with the new cluster’s kubeconfig to install the provider
>  components.
> 
>     Example:
> 
>     `clusterctl init --kubeconfig target.yaml --infrastructure metal3:v1.1.0
>   --target-namespace metal3 --watching-namespace metal3`
> 

Reviewer 4: ```suggestion
7. Use `clusterctl` init with the new cluster’s kubeconfig to install the provider
 components.
    ```bash
     clusterctl init --kubeconfig target.yaml --infrastructure metal3:v1.1.0
    --target-namespace metal3 --watching-namespace metal3
    ```
```

> 8. Use `clusterctl` move to move the Cluster API resources from the bootstrap
>  cluster to the target management cluster.
> 
>     Example:
> 
>     `clusterctl move --to-kubeconfig target.yaml -n metal3 -v 10`
> 

Reviewer 4: ```suggestion
8. Use `clusterctl` move to move the Cluster API resources from the bootstrap
 cluster to the target management cluster.
 
    ```bash
    clusterctl move --to-kubeconfig target.yaml -n metal3 -v 10
    ```

```

# CAPM3 Pivoting

## What is pivoting

Cluster API Provider Metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.

CAPI Pivoting feature is a process of moving the provider components and declared Cluster API resources from a source
management cluster to a target management cluster by using the `clusterctl` functionality called "move".
More information about the general CAPI "move" functionality can be found [here](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html).

In Metal3, pivoting is performed by using the CAPI `clusterctl` tool provided by Cluster-API project. `clusterctl` recognizes pivoting as move.
During the pivot process `clusterctl` pauses any reconciliation of CAPI objects and this gets propagated to CAPM3 objects as well.
Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
bootstrap cluster.

## Prerequisite

> * **It is mandatory to use `clusterctl` for both the bootstrap and target cluster.**
> 
>   If the provider components are not installed using `clusterctl`, it will not
>   be able to identify the objects to move.  Initializing the cluster using
>   `clusterctl` essentially adds the following labels in the CRDs of each related object.
> 
>   ```yaml
>   labels:
>   - clusterctl.cluster.x-k8s.io: ""
>   - cluster.x-k8s.io/provider: "<provider-name>"
>   ```
> 
>   So if the clusters are not initialized using `clusterctl`, all the CRDS of the
>   objects to be moved to target cluster needs to have these labels both in
>   bootstrap cluster and target cluster before performing the move.
> 
>   **Note**: This is not recommended, since
>   the way `clusterctl` identifies objects to manage might change in the future, so
>   it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.
> 
> * **BareMetalHost objects have correct status annotation.**
> 
>   Since BareMetalHost (BMH) _status_ holds important information regarding the BMH itself, BMH with
>   status has to be moved and it has to be reconstructed with correct status in
>   target  cluster before it is being reconciled. This is now done through BMH
>   status annotation in [BMO](../bmo/introduction.md).
> 
> * **Maintain connectivity towards provisioning network.**
> 
>   Baremetal machines boot over a network with a DHCP server. This requires
>   maintaining a fixed IP end points towards the provisioning network. This is
>   achieved through _keepalived_. A new container is added namely
>   _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
>   keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
>   is to ensure that the Ironic Endpoint IP is also passed onto the target
>   cluster control plane. This also guarantees that once moving is done and the
>   management cluster is taken down, target cluster controlplane can re-claim
>   the Ironic endpoint IP through keepalived. The end goal is to make Ironic
>   endpoint reachable in the target cluster.
> 
> * **BMO is deployed as part of CAPM3.**
> 
>   If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
>   need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
>   because since CAPM3 release [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
>   BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
>   This is a prerequisite for both the management and the target cluster.
> 
> * **Objects should have a proper owner reference chain.**
> 
>   `clusterctl move` moves all the objects to the target cluster following the
> [owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
> So, it is necessary to verify that all the desired objects that needs to
> be moved to the target cluster have a proper owner reference chain.
> 

Reviewer 4: ```suggestion
1. **It is mandatory to use `clusterctl` for both the bootstrap and target cluster**

      If the provider components are not installed using `clusterctl`, it will not
      be able to identify the objects to move.  Initializing the cluster using
      `clusterctl` essentially adds the following labels in the CRDs of each related object.

      ```yaml
      labels:
      - clusterctl.cluster.x-k8s.io: ""
      - cluster.x-k8s.io/provider: "<provider-name>"
      ```

      So if the clusters are not initialized using `clusterctl`, all the CRDS of the
      objects to be moved to target cluster needs to have these labels both in
      bootstrap cluster and target cluster before performing the move.

      **Note**: This is not recommended, since
      the way `clusterctl` identifies objects to manage might change in the future, so
      it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.

1. **BareMetalHost objects have correct status annotation**

   Since BareMetalHost (BMH) _status_ holds important information regarding the BMH itself, BMH with
   status has to be moved and it has to be reconstructed with correct status in
   target  cluster before it is being reconciled. This is now done through BMH
   status annotation in [BMO](../bmo/introduction.md).

1. **Maintain connectivity towards provisioning network**

   Baremetal machines boot over a network with a DHCP server. This requires
   maintaining a fixed IP end points towards the provisioning network. This is
   achieved through _keepalived_. A new container is added namely
   _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
   keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
   is to ensure that the Ironic Endpoint IP is also passed onto the target
   cluster control plane. This also guarantees that once moving is done and the
   management cluster is taken down, target cluster controlplane can re-claim
   the Ironic endpoint IP through keepalived. The end goal is to make Ironic
   endpoint reachable in the target cluster.

1. **BMO is deployed as part of CAPM3**

   If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
   need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
   because since CAPM3 release [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
   BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
   This is a prerequisite for both the management and the target cluster.

1. **Objects should have a proper owner reference chain**

   `clusterctl move` moves all the objects to the target cluster following the
   [owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
   So, it is necessary to verify that all the desired objects that needs to
   be moved to the target cluster have a proper owner reference chain.
```

Reviewer 4: so that it looks like
![pre](https://user-images.githubusercontent.com/35802557/159922022-588ac489-4280-4c83-820a-b22134f27991.png)

Reviewer 4: and probably should be **Prerequisites** (in plural)

Author: It was originally like that but then I was informed to change it to Prerequisite

> * **It is mandatory to use `clusterctl` for both the bootstrap and target cluster.**
> 
>   If the provider components are not installed using `clusterctl`, it will not
>   be able to identify the objects to move.  Initializing the cluster using
>   `clusterctl` essentially adds the following labels in the CRDs of each related object.
> 
>   ```yaml
>   labels:
>   - clusterctl.cluster.x-k8s.io: ""
>   - cluster.x-k8s.io/provider: "<provider-name>"
>   ```
> 
>   So if the clusters are not initialized using `clusterctl`, all the CRDS of the
>   objects to be moved to target cluster needs to have these labels both in
>   bootstrap cluster and target cluster before performing the move.
> 
>   **Note**: This is not recommended, since
>   the way `clusterctl` identifies objects to manage might change in the future, so
>   it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.
> 
> * **BareMetalHost objects have correct status annotation.**
> 
>   Since BareMetalHost (BMH) _status_ holds important information regarding the BMH itself, BMH with
>   status has to be moved and it has to be reconstructed with correct status in
>   target  cluster before it is being reconciled. This is now done through BMH
>   status annotation in [BMO](../bmo/introduction.md).
> 
> * **Maintain connectivity towards provisioning network.**
> 
>   Baremetal machines boot over a network with a DHCP server. This requires
>   maintaining a fixed IP end points towards the provisioning network. This is
>   achieved through _keepalived_. A new container is added namely
>   _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
>   keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
>   is to ensure that the Ironic Endpoint IP is also passed onto the target
>   cluster control plane. This also guarantees that once moving is done and the
>   management cluster is taken down, target cluster controlplane can re-claim
>   the Ironic endpoint IP through keepalived. The end goal is to make Ironic
>   endpoint reachable in the target cluster.
> 
> * **BMO is deployed as part of CAPM3.**
> 
>   If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
>   need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
>   because since CAPM3 release [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
>   BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
>   This is a prerequisite for both the management and the target cluster.
> 
> * **Objects should have a proper owner reference chain.**
> 
>   `clusterctl move` moves all the objects to the target cluster following the
> [owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
> So, it is necessary to verify that all the desired objects that needs to
> be moved to the target cluster have a proper owner reference chain.
> 

Reviewer 4: ```suggestion
1. **It is mandatory to use `clusterctl` for both the bootstrap and target cluster**

      If the provider components are not installed using `clusterctl`, it will not
      be able to identify the objects to move.  Initializing the cluster using
      `clusterctl` essentially adds the following labels in the CRDs of each related object.

      ```yaml
      labels:
      - clusterctl.cluster.x-k8s.io: ""
      - cluster.x-k8s.io/provider: "<provider-name>"
      ```

      So if the clusters are not initialized using `clusterctl`, all the CRDS of the
      objects to be moved to target cluster needs to have these labels both in
      bootstrap cluster and target cluster before performing the move.

      **Note**: This is not recommended, since
      the way `clusterctl` identifies objects to manage might change in the future, so
      it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.

1. **BareMetalHost objects have correct status annotation**

   Since BareMetalHost (BMH) _status_ holds important information regarding the BMH itself, BMH with
   status has to be moved and it has to be reconstructed with correct status in
   target  cluster before it is being reconciled. This is now done through BMH
   status annotation in [BMO](../bmo/introduction.md).

1. **Maintain connectivity towards provisioning network**

   Baremetal machines boot over a network with a DHCP server. This requires
   maintaining a fixed IP end points towards the provisioning network. This is
   achieved through _keepalived_. A new container is added namely
   _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
   keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
   is to ensure that the Ironic Endpoint IP is also passed onto the target
   cluster control plane. This also guarantees that once moving is done and the
   management cluster is taken down, target cluster controlplane can re-claim
   the Ironic endpoint IP through keepalived. The end goal is to make Ironic
   endpoint reachable in the target cluster.

1. **BMO is deployed as part of CAPM3**

   If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
   need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
   because since CAPM3 release [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
   BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
   This is a prerequisite for both the management and the target cluster.

1. **Objects should have a proper owner reference chain**

   `clusterctl move` moves all the objects to the target cluster following the
   [owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
   So, it is necessary to verify that all the desired objects that needs to
   be moved to the target cluster have a proper owner reference chain.
```

# CAPM3 Pivoting

## What is pivoting

> CAPM3 implements support for CAPI's 'move/pivoting' feature.

Reviewer 4: ```suggestion
Cluster-api-providfer-metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.
```

> Cluster-api-providfer-metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.

Author: yes it is intentional.

> CAPI Pivoting feature is a process for moving the provider components and declared Cluster API resources from a source

Reviewer 4: ```suggestion
CAPI Pivoting feature is a process of moving the provider components and declared Cluster API resources from a source
```

Author: done

management cluster to a target management cluster by using the clusterctl functionality called "move".
More information about the general CAPI "move" functionality can be find [here](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html).

In Metal3, pivoting is performed by using the CAPI clusterctl tool provided by Cluster-API. clusterctl recognizes pivoting as move.

> During the pivot process clusterctl pauses any reconciliation of Cluster-API objects and this gets propagated to Cluster-api-provider-metal3
> (CAPM3) objects as well. Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the

Reviewer 4: ```suggestion
During the pivot process clusterctl pauses any reconciliation of Cluster-API objects and this gets propagated to
CAPM3 objects as well. Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
```

> Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
> ephemeral cluster.

Reviewer 4: ```suggestion
Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
source cluster.
```
just to use the same terminology

Author: done

## Prerequisits

> * It is mandatory to use **`clusterctl` both the bootstrap and target cluster.**

Reviewer 4: This sentence seems a bit incomplete. Also
```suggestion
* It is mandatory to use `clusterctl` both the bootstrap and target cluster.
```

> * **It is mandatory to use `clusterctl` both the bootstrap and target cluster.**

Reviewer 4: This sentence seems missing something

Author: done

  If the provider components are not installed using `clusterctl`, it will not

>   be able to identify the objects to move.  Initializing the cluster using

Reviewer 4: Since it is hard to show what changes are requested, I will mark them in both.

So if the clusters are not initialized using clusterctl, all the CRDS of the objects to be moved to target cluster need to have these labels both **before performing move**. Note: **this** is not recommended, since the way clusterctl identifies objects to manage might change in future, so it’s always safe to install CRDs and controllers through **`clusterctl init`**.

Author: done

3. Use clusterctl init to install the provider components

   Example:

>    `clusterctl init --infrastructure metal3:v1.1.
>    --target-namespace metal3 --watching-namespace metal3`

Reviewer 4: ```suggestion
   `clusterctl init --infrastructure metal3:v1.1.1
   --target-namespace metal3 --watching-namespace metal3`
```

Author: done

>  This command will create the necessary CAPI controllers (CAPI, CABPK, CAKCP)
>  and CAPM3 as the infrastructure provider which also includes the Baremetal
>  Operator (BMO). All of the controllers will be installed on namespace `metal3`
>  and they will be watching over objects in namespace `metal3`.

Reviewer 4: ```suggestion
 This command will create the necessary CAPI controllers (CAPI, CABPK, CAKCP)
 and CAPM3 as the infrastructure provider. All of the controllers will be installed on namespace `metal3`
 and they will be watching over objects in namespace `metal3`.
```
BMO is not part of CAPM3 anymore

Author: done

   In `metal3-dev-env` steps 1 and 2 are part of `make` command.
4. Provision target cluster:

   Example:

   `clusterctl config cluster ... | kubectl apply -f -`

   In `metal3-dev-env` this is done through the scripts in `scripts/v1alphaX/`
 directory.

5. Wait for the target management cluster to be up and running and once it is up
 get the kubeconfig for the new target management cluster.

6. Use the new cluster’s kubeconfig to install the ironic-components in the
 target cluster.

7. Use clusterctl init with the new cluster’s kubeconfig to install the provider

>  components.

Reviewer 4: I feel like we don't need to give examples with dev-env, but just generic instructions on how to perform move.

Author: done

Author: This has to stay like this because otherwise markdownlint has issue with it.
